indexer/lib/service-provider-registry-handlers.js

```diff
@@ -0,0 +1,283 @@
+      productType,
+    })
+    return new Response('Bad Request', { status: 400 })
+  }
+  if (Number(productType) !== PRODUCT_TYPE_PDP) {
+    return new Response('OK', { status: 200 })
+  }
+
+  return await handleProviderServiceUrlUpdate(
+    env,
+    rpcRequest,
+    providerId,
+    rpcUrl,
+    glifToken,
+    blockNumber,
+    serviceProviderRegistryAddress,
+  )
+}
+
+/**
+ * @param {Env} env
+ * @param {function} rpcRequest
+ * @param {string | number} providerId
+ * @param {string | number} productType
+ * @param {string} rpcUrl
+ * @param {string} glifToken
+ * @param {number | 'latest' | 'earliest' | 'pending'} blockNumber
+ * @param {string} serviceProviderRegistryAddress
+ * @returns {Promise<Response>}
+ */
+export async function handleProductUpdated(
+  env,
+  rpcRequest,
+  providerId,
+  productType,
+  rpcUrl,
+  glifToken,
+  blockNumber,
+  serviceProviderRegistryAddress,
+) {
+  if (
+    (typeof providerId !== 'string' && typeof providerId !== 'number') ||
+    (typeof productType !== 'string' && typeof productType !== 'number')
+  ) {
+    console.error('ServiceProviderRegistry.ProductUpdated: Invalid payload', {
+      providerId,
+      productType,
+    })
+    return new Response('Bad Request', { status: 400 })
+  }
+  if (Number(productType) !== PRODUCT_TYPE_PDP) {
+    return new Response('OK', { status: 200 })
+  }
+
+  return await handleProviderServiceUrlUpdate(
+    env,
+    rpcRequest,
+    providerId,
+    rpcUrl,
+    glifToken,
+    blockNumber,
+    serviceProviderRegistryAddress,
+  )
+}
+
+/**
+ * @param {Env} env
+ * @param {string | number} providerId
+ * @param {string | number} productType
+ * @returns {Promise<Response>}
+ */
+export async function handleProductRemoved(env, providerId, productType) {
+  if (
+    (typeof providerId !== 'string' && typeof providerId !== 'number') ||
+    (typeof productType !== 'string' && typeof productType !== 'number')
+  ) {
+    console.error('ServiceProviderRegistry.ProductRemoved: Invalid payload', {
+      providerId,
+      productType,
+    })
+    return new Response('Bad Request', { status: 400 })
+  }
+  if (Number(productType) !== PRODUCT_TYPE_PDP) {
+    return new Response('OK', { status: 200 })
+  }
+
+  const result = await env.DB.prepare(
+    `
+        DELETE FROM providers WHERE id = ?
+      `,
+  )
+    .bind(providerId)
+    .run()
+  if (result.meta.changes === 0) {
+    return new Response('Provider Not Found', { status: 404 })
+  }
+  return new Response('OK', { status: 200 })
+}
+
+/**
+ * @param {Env} env
+ * @param {string | number} providerId
+ * @returns {Promise<Response>}
+ */
+export async function handleProviderRemoved(env, providerId) {
+  if (typeof providerId !== 'string' && typeof providerId !== 'number') {
+    console.error('ServiceProviderRegistry.ProviderRemoved: Invalid payload', {
+      providerId,
+    })
+    return new Response('Bad Request', { status: 400 })
+  }
+
+  const result = await env.DB.prepare(
```

Comment:
This might cause a edge-case that @bajtos has mentioned [above](https://github.com/filcdn/worker/pull/224/files#r2298249579). 

If provider deletes the product before it gets deleted from the registry we'll get response with status code `404`.

Reply:
could you please describe this edge case more?